### PR TITLE
fix(gatsby-transformer-remark): fix excerpt generation - strip excessive white spaces, extract alt from images

### PR DESCRIPTION
**Pull request by macklinu**
## Description

This is a failing test case to reproduce the issue stated in https://github.com/gatsbyjs/gatsby/issues/12398. Here is a screenshot of test failure output:

image

## Related Issues

Relates to #12398

### CODE REVIEW

**Comment by wardpeet** [comment]
@macklinu mind fixing this as well? Thanks for creating the test, if you don't really know how we can take over from here.

**Comment by macklinu** [comment]
@wardpeet I'll take a look at patching this and ping you if I come up with a solution or need to ask for another contributor's help. Thank you!

**Comment by macklinu** [comment]
Would it make sense to use [`strip-markdown`](https://github.com/remarkjs/strip-markdown) as a remark plugin instead of writing our own markdown processor in order to parse out the excerpt content? Or maybe we don't need all the functionality of `strip-markdown` and just need to account for a few more node types? Any implementation suggestions?

https://github.com/gatsbyjs/gatsby/blob/15ef1783620154abefdd5eac9e8d5145ded68d09/packages/gatsby-transformer-remark/src/extend-node-type.js#L412-L449

**Comment by wardpeet** [comment]
strip-markdown strips markdown away, we still want to show the link or bold text right? That plugin just makes the text unstyled.

**Comment by macklinu** [comment]
Ah, I see. I think my failing test assertion is incorrect then. We don't want to necessarily strip markdown, but we just want to preserve markdown as is and not remove any content?

**Comment by wardpeet** [comment]
yes, the issue is that an extra space is added between the link & the ! symbol. It might be something that we have to fix upstream.

**Comment by frinyvonnick** [comment]
Hi @macklinu 👋, 

First thanks for your failing test it really helped me diving in this bug 👏. 

I found out that the function `getExcerpt` in [extend-node-type.js](https://github.com/gatsbyjs/gatsby/pull/12878/commits/f6a6e33388adba13d8612496011e8891e865b146#diff-488d99b036db541ee4a25a6991027b5dR412) wasn't handling `image` and the `link` was classified as an `image`. That was causing the missing parts. The second problem of extra spaces was caused by a join of excerpt parts with a space. 

Sorry to not asking before pushing on your branch 😊

File: packages/gatsby-transformer-remark/src/__tests__/extend-node.js [hunk 1]
```diff
@@ -376,6 +376,27 @@ Where oh [*where*](nick.com) **_is_** ![that pony](pony.png)?`,
...
+
+Where oh [*where*](nick.com) **_is_** ![that pony](pony.png)?`,
+    `excerpt
+      excerptAst
+      frontmatter {
+          title
+      }
+      `,
+    node => {
+      expect(node).toMatchSnapshot()
```
Comment (pieh) [inline packages/gatsby-transformer-remark/src/__tests__/extend-node.js]: Do we need this assertion?

File: packages/gatsby-transformer-remark/src/__tests__/extend-node.js [hunk 2]
```diff
@@ -376,6 +376,27 @@ Where oh [*where*](nick.com) **_is_** ![that pony](pony.png)?`,
...
+  bootstrapTest(
+    `excerpt does have missing words and extra spaces`,
+    `---
+title: "my little pony"
+date: "2017-09-18T23:19:51.246Z"
+---
+
+Where oh [*where*](nick.com) **_is_** ![that pony](pony.png)?`,
+    `excerpt
+      excerptAst
```
Comment (pieh) [inline packages/gatsby-transformer-remark/src/__tests__/extend-node.js]: I would say this is not needed (if we remove snapshot for this test)

**Comment by frinyvonnick** [comment]
Thanks for your review @pieh 👍

**Comment by wardpeet** [comment]
Sweet! thanks for getting this PR through the door @frinyvonnick! I'll fix the unit test.

**Comment by frinyvonnick** [comment]
Thanks @wardpeet 🙏 !

**Comment by wardpeet** [comment]
pretty sure we can ignore danger here.

**Comment by sidharthachatterjee** [comment]
Published in `gatsby-transformer-remark@2.3.12`

**Pull request merged**